sys/unix/nleshared.cc

```diff
@@ -0,0 +1,798 @@
+    }
+  }
+  size_t fileSize() const {
+    struct ::stat s;
+    if (::fstat(fd, &s)) {
+      throw std::system_error(errno, std::system_category(), "fstat");
+    }
+    return s.st_size;
+  }
+};
+
+struct Mmap {
+  std::byte* data_ = nullptr;
+  size_t size_ = 0;
+  Mmap() = default;
+  Mmap(std::byte* data, size_t size) noexcept : data_(data), size_(size) {}
+  Mmap(const Mmap&) = delete;
+  Mmap(Mmap&& n) noexcept {
+    data_ = std::exchange(n.data_, nullptr);
+    size_ = n.size_;
+  }
+  Mmap& operator=(const Mmap&) = delete;
+  Mmap& operator=(Mmap&& n) noexcept {
+    std::swap(data_, n.data_);
+    std::swap(size_, n.size_);
+    return *this;
+  }
+  ~Mmap() {
+    if (data_) {
+      ::munmap(data_, size_);
+    }
+  }
+  static Mmap make(const Fd& fd, size_t offset, size_t size, bool write, bool shared) {
+    errno = 0;
+    void* ptr = ::mmap(nullptr, size, write ? PROT_READ | PROT_WRITE : PROT_READ, shared ? MAP_SHARED : MAP_PRIVATE, (int)fd, offset);
+    if (!ptr || ptr == MAP_FAILED) {
+      throw std::system_error(errno, std::system_category(), "mmap");
+    }
+    return {(std::byte*)ptr, size};
+  }
+  static Mmap shared(const Fd& fd, size_t offset, size_t size, bool write) {
+    return make(fd, offset, size, write, true);
+  }
+  static Mmap private_(const Fd& fd, size_t offset, size_t size, bool write) {
+    return make(fd, offset, size, write, false);
+  }
+  operator bool() const noexcept {
+    return data_;
+  }
+  operator std::byte*() const noexcept {
+    return data_;
+  }
+  std::byte* data() noexcept {
+    return data_;
+  }
+  const std::byte* data() const noexcept {
+    return data_;
+  }
+  size_t size() const noexcept {
+    return size_;
+  }
+};
+
+struct RelativeRelocation {
+  uint64_t targetOffset;
+  uint64_t valueOffset;
+};
+
+struct Image {
+  Mmap mem;
+  std::byte* base = nullptr;
+
+  operator bool() const noexcept {
+    return mem;
+  }
+};
+
+class NHLoader {
+  // The raw file data is kept around as ELF headers are generally read from it,
+  // and there is no guarantee that the in-memory layout mirrors the file data.
+  Mmap fileMem;
+  Elf64_Ehdr* hdr;
+  // The "base" form of the library, mapped into memory with all symbols resolved.
+  Fd sharedFd;
+  Image sharedImage;
+  // Relocations that must be performed at fork time
+  std::vector<RelativeRelocation> forkRelocations;
+  // Relocations that must be performed at reset time
+  std::vector<RelativeRelocation> resetRelocations;
+
+  size_t initfunc;
+  size_t finifunc;
+  size_t initarray;
+  size_t initarraysz;
+  size_t finiarray;
+  size_t finiarraysz;
+
+  void load(std::string libPath, const std::unordered_map<std::string, void*>& overrides) {
+    Fd diskFd = open(libPath.c_str(), O_RDONLY);
+
+    fileMem = Mmap::shared(diskFd, 0, diskFd.fileSize(), false);
+
+    hdr = (Elf64_Ehdr*)fileMem.data();
+    boundsCheck(hdr);
+
+    size_t baseaddr = ~0;
+    size_t endaddr = 0;
+
+    forPh([&](Elf64_Phdr* ph) {
```

Comment:
I'm assuming this is fast enough to not need the ability to `break` based on e.g. a return value of the lambda?

Reply:
This code only runs once in the entire lifespan of the process, so... There's really nothing to be gained by making this stuff faster.